src/main/java/bio/terra/service/load/LoadDao.java

```diff
@@ -72,83 +78,94 @@ public LoadDao(NamedParameterJdbcTemplate jdbcTemplate) {
    *
    * @param loadTag tag identifying this load
    * @param flightId flight id taking the lock
+   * @param datasetId dataset to which this load is being applied
    * @return Load object including the load id
    */
   @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
-  public Load lockLoad(String loadTag, String flightId) throws InterruptedException {
+  public Load lockLoad(String loadTag, String flightId, UUID datasetId) {
     jdbcTemplate.getJdbcTemplate().execute("LOCK TABLE load IN EXCLUSIVE MODE");
 
     String upsert =
-        "INSERT INTO load (load_tag, locked, locking_flight_id)"
-            + " VALUES (:load_tag, true, :flight_id)"
-            + " ON CONFLICT ON CONSTRAINT load_load_tag_key DO NOTHING";
+        """
+        INSERT INTO load (load_tag, locked, locking_flight_id, dataset_id)
+        VALUES (:load_tag, true, :locking_flight_id, :dataset_id)
+        ON CONFLICT ON CONSTRAINT load_load_tag_dataset_id_key DO NOTHING
+        """;
 
     MapSqlParameterSource params =
-        new MapSqlParameterSource().addValue("load_tag", loadTag).addValue("flight_id", flightId);
+        new MapSqlParameterSource()
+            .addValue(LOAD_TAG, loadTag)
+            .addValue(LOCKING_FLIGHT_ID, flightId)
+            .addValue(DATASET_ID, datasetId);
     DaoKeyHolder keyHolder = new DaoKeyHolder();
-    int rows = jdbcTemplate.update(upsert, params, keyHolder);
-    Load load;
-    if (rows == 0) {
-      // We did not insert. Therefore, someone has the load tag locked.
-      // Retrieve it, in case it is us re-locking
-      load = lookupLoadByTag(loadTag);
-      if (load == null) {
-        throw new CorruptMetadataException("Load row should exist! Load tag: " + loadTag);
-      }
-
-      // It is locked by someone else
-      if (!StringUtils.equals(load.getLockingFlightId(), flightId)) {
-        throw new LoadLockedException(
-            "Load '" + loadTag + "' is locked by flight '" + load.getLockingFlightId() + "'");
-      }
-    } else {
-      load =
-          new Load()
-              .id(keyHolder.getId())
-              .loadTag(keyHolder.getString("load_tag"))
-              .locked(keyHolder.getField("locked", Boolean.class).orElse(false))
-              .lockingFlightId(keyHolder.getString("locking_flight_id"));
+    jdbcTemplate.update(upsert, params, keyHolder);
```

Comment:
Since `keyHolder` is now unused, it can be removed.

Or, you can preserve the existing logic by building a `Load` object if the `update()` command returns a nonzero value. The only field besides the key that's db generated is the date, which could be generated in this method instead of in the database.